### PR TITLE
Encapsulation rework for simplified multi-node operations

### PR DESCRIPTION
**Pull request by tgraf**
This  is a major rework of the encapsulation mode and drastically simplifies operations:
 - No longer requires any network configuration, the IP of the VM/host is automatically used as tunnel endpoint across the mesh
 - There is no longer a need to configure any routes for the container prefixes in the cloud network or the underlying fabric
 - The node prefix to node ip mapping is automatically derived from the Kubernetes PodCIDR

### CODE REVIEW

File: pkg/bpf/map.go
Comment (aanm) [inline pkg/bpf/map.go:114]: Can you write the name of 1 or 2 maps that don't contain persistent data?

File: pkg/maps/tunnel/tunnel.go
Comment (aanm) [inline pkg/maps/tunnel/tunnel.go]: `// represents both IPv6 and IPv4`?

File: pkg/maps/tunnel/tunnel.go
Comment (aanm) [inline pkg/maps/tunnel/tunnel.go]: Since this function is likely called first, can we put this function on top?

File: contrib/vagrant/start.sh
Comment (aanm) [inline contrib/vagrant/start.sh]: I'm assuming `10.x.0.1/16` will work as fine, right?

File: daemon/daemon.go
Comment (aanm) [inline daemon/daemon.go:469]: Do you want me to replace `nodeaddress.SetNodeAddress(ipv6NodeAddress, ciliumIPv4.NodeIP().String(), "")` in https://github.com/cilium/cilium/pull/1039/commits/159237e067588cd425c2169e74bc95550d33ccaf with this?

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: Out of curiosity, when is this less than 4?

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go:183]: Why not get rid of the Getters and Setters and access the IPs/IPNets directly?

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: should be `len(raw) != 16`, also be aware an `net.IPv4` address has also len 16

File: tests/manual-multi-node.bash
Comment (aanm) [inline tests/manual-multi-node.bash:70]: Change the `daemon/main.go`
It is currently as this:
```
IPv4 allocation range for local endpoints, must be in correct format (10.N.0.1/16)
```

File: daemon/daemon.go
Comment (aanm) [inline daemon/daemon.go]: Why not `common.EndpointsPerHost`? With a queue of 512 There can be potentials deadlocks if you have >=512 endpoints and regenerate all endpoints at the same time

File: pkg/maps/lxcmap/lxcmap.go
Comment (aanm) [inline pkg/maps/lxcmap/lxcmap.go]: Init on top?

File: daemon/main.go
Comment (aanm) [inline daemon/main.go]: If the default is auto why not set it up in the default value?
```
"ipv4-range", "auto", "Per-node IPv4 endpoint prefix")
```
Same with `ipv6-range`

File: daemon/main.go
Comment (aanm) [inline daemon/main.go]: `s/(default; auto)/(default: auto)`

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: We have a mask for this, no? `StateIPv6Mask`

File: daemon/main.go
Comment (aanm) [inline daemon/main.go]: The same `auto` comments I've made in the previous commit.

Any constrains with this `node-ipv6`? Can be anything I want? Won't cilium fail if I assign a link-local `fe80` on purpose?

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: Did you forget to copy the IP? Before changing it?

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: Did you forget to copy the IP? Before changing it?

File: daemon/daemon.go
Comment (tgraf) [reply daemon/daemon.go]: If this is the code then the surrounding code needs to be changed. We can have the same endpoint being rebuilt multiple times. I've bumped endpoints per host to 64K. We absolutely don't want a build backlog of 64K.

File: daemon/daemon.go
Comment (tgraf) [reply daemon/daemon.go:469]: Yes, 159237e should be rebased on top of this.

File: tests/manual-multi-node.bash
Comment (tgraf) [reply tests/manual-multi-node.bash:70]: That change was in a later commit, I moved it to earlier now.

File: daemon/daemon.go
Comment (aanm) [reply daemon/daemon.go]: >We can have the same endpoint being rebuilt multiple times.

We can but, for example, if you try to regenerate the same endpoint 100 times in 1 second, the queue will only contain a single entry of that endpoint, not 100 entries.
The entries in the queue are guarantee to be unique.

Comment (tgraf) [reply daemon/daemon.go]: Discussed a bit on slack. I've put the original channel size in place again. We will look at this issue in a separate PR.

File: pkg/nodeaddress/node_address.go
Comment (tgraf) [reply pkg/nodeaddress/node_address.go:183]: I had direct access at first but then figured that having Get and Set function allows to do validation as is now done for some and it allows to lock the variables if needed later on.

I didn't really see a downside of using Getters and Setters here.

File: daemon/main.go
Comment (tgraf) [reply daemon/main.go]: Interesting idea. I like it.

File: daemon/main.go
Comment (tgraf) [reply daemon/main.go]: Good point regarding link local. I'll check for this.

File: pkg/bpf/map.go
Comment (aanm) [reply pkg/bpf/map.go:114]: not addressed?

File: pkg/node/manager.go
Comment (aanm) [inline pkg/node/manager.go:46]: Small nit, why not throwing this error to the caller?

File: pkg/node/manager.go
Comment (aanm) [inline pkg/node/manager.go]: Didn't we discuss that we wouldn't delete the nodes from the golang map in case the bpf throw an error?

File: contrib/vagrant/start.sh
Comment (aanm) [reply contrib/vagrant/start.sh]: addressed?

File: daemon/daemon.go
Comment (aanm) [inline daemon/daemon.go]: Should we ignore this? It's the router address where a container can potentially colide with this one, won't it will be able to receive all traffic from the remaining endpoints?

File: daemon/daemon.go
Comment (aanm) [inline daemon/daemon.go]: small nit, replace `"auto"` with a const?

File: daemon/main.go
Comment (aanm) [inline daemon/main.go]: Please write an example of the expected `ipv4-range`.

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: small nit:
```
if err != nil{
    log.Debugf("error...")
    return
}
// rest of the code
```

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: Don't we have a function that sets the IPv6 address with `ffff` at the end?

File: pkg/nodeaddress/node_address.go
Comment (aanm) [reply pkg/nodeaddress/node_address.go]: addressed?

File: pkg/nodeaddress/node_address.go
Comment (aanm) [inline pkg/nodeaddress/node_address.go]: I think what you want is 
```
range4, range6 := *ipv4AllocRange, *ipv6AllocRange
```
and on line 224
```
		IPv4AllocCIDR: &range4,
		IPv6AllocCIDR: &range6,
```

otherwise if you change `ipv4AllocRange` in the future, the value is also changed in `range4`

File: daemon/daemon.go
Comment (aanm) [inline daemon/daemon.go:130]: We don't need the lxcmap anymore?

File: pkg/bpf/map.go
Comment (aanm) [inline pkg/bpf/map.go:220]: I'm just writing this as a precaution: between line 220 and 306 are any garbage of maps file descriptors left open?

File: pkg/bpf/map.go
Comment (aanm) [inline pkg/bpf/map.go]: my previous comment is regarding this `goto`, the `fd` open on the next line doesn't seem to be properly closed (I think)

File: pkg/maps/lxcmap/lxcmap.go
Comment (aanm) [inline pkg/maps/lxcmap/lxcmap.go]: no need for `log.Fatal` here?

File: pkg/maps/lxcmap/lxcmap.go
Comment (aanm) [inline pkg/maps/lxcmap/lxcmap.go]: `FIXME: revert previous updates?`

File: daemon/daemon.go
Comment (aanm) [inline daemon/daemon.go]: replace `"auto"` with const?

File: daemon/main.go
Comment (aanm) [inline daemon/main.go]: What was the outcome of running it with a link-local ipv6 address does cilium handle it well?

also:
`"Per-node IPv6 endpoint prefix, must be /%d", IPv6NodePrefixLen`?

File: daemon/main.go
Comment (aanm) [inline daemon/main.go]: I would like to see an example of what the user should write here

File: daemon/main.go
Comment (aanm) [inline daemon/main.go]: I would like to see an example of what the user should write here

File: daemon/daemon.go
Comment (aanm) [inline daemon/daemon.go:338]: this is a workaround right?

File: tests/k8s/multi-node/cluster/cluster-manager.bash
Comment (aanm) [inline tests/k8s/multi-node/cluster/cluster-manager.bash]: 2 cilium-lb? The cilium-lb is a daemon-set to test k8s ingress with a specific node selector label. There should be a single ingress controller running in the k8s cluster otherwise we will end up with conflicting IPs when trying to update the frontend IP in the kubernetes ingress resource.

**Review by borkmann — APPROVED** [review]
LGTM, would be great if we could get rid of masquerade target at some point later and also do this in BPF.

File: pkg/bpf/map.go
Comment (ianvernon) [reply pkg/bpf/map.go:220]: Nit: function name is a bit misleading; I would expect 'migrate' to migrate data from an old map to a new map, but this function checks for differences between the existing map information in the BPF FS and the information Cilium contains about said map without migrating any data between the old map and the new one. Maybe rename to 'resolveMapMismatch' or something more evocative of what the function does.

**Review by ianvernon — APPROVED** [review]
Reviewed the Go code - looks great. Have one small nitpick which does not need a second review.

File: pkg/bpf/map.go
Comment (tgraf) [reply pkg/bpf/map.go:114]: `pkg/bpf/map` is generic and not cilium specific at all. It is already being used by projects outside of Cilium. Using Cilium specific examples would be confusing. Is the description not clear?

File: pkg/bpf/map.go
Comment (tgraf) [reply pkg/bpf/map.go:220]: In the future, migrate can actually perform migration. We are not maintaining backwards compatibility yet though. Still confusing?

File: pkg/node/manager.go
Comment (tgraf) [reply pkg/node/manager.go:46]: How do you want to handle the error?

File: pkg/node/manager.go
Comment (tgraf) [reply pkg/node/manager.go]: Agreed. I think this should also use the state mechanism once we have it which then attempts to delete multiple times. I will change this to keep the node for now.

File: daemon/daemon.go
Comment (tgraf) [reply daemon/daemon.go:130]: lxcmap instance is now maintained in the package itself. There is always only a single instance.

File: daemon/daemon.go
Comment (tgraf) [reply daemon/daemon.go:338]: No. Do you have a better idea?

File: pkg/bpf/map.go
Comment (aanm) [reply pkg/bpf/map.go:114]: Yes, I thought this was cilium specific and that you were removing maps on startup. If it's generic I'm good.

File: pkg/node/manager.go
Comment (aanm) [reply pkg/node/manager.go:46]: I was thinking on what we discussed on keeping a cache, this is fine then

**Comment by aanm** [comment]
Regarding the masquerade I'm not currently seeing why we need to use iptables I need to check with you what the exact issue was. But I'm with @borkmann if we could get rid of masquerade target at some point later and also do this in BPF it would be great. I'm good as it is right now.

File: pkg/node/manager.go
Comment (tgraf) [reply pkg/node/manager.go:46]: OK, understood. We can add it when we need it.

**Comment by tgraf** [comment]
> Regarding the masquerade I'm not currently seeing why we need to use iptables I need to check with you what the exact issue was. But I'm with @borkmann if we could get rid of masquerade target at some point later and also do this in BPF it would be great. I'm good as it is right now.

The issue is this:

- When a service loadbalances to external IPs (a typical example is the kubernetes API server service), cilium will rewrite the destination IP to that external IP and then give the packet to Linux for routing. The source IP is the one of the container. The container IP is not routable if we are running in encapsulation mode

- When talking to external IPs outside of the cluster, masquerading is required anyway. This can be done by the cloud provider in which case users would run with masquerading turned off.

**Comment by aanm** [comment]
>>Regarding the masquerade I'm not currently seeing why we need to use iptables I need to check with you what the exact issue was. But I'm with @borkmann if we could get rid of masquerade target at some point later and also do this in BPF it would be great. I'm good as it is right now.
>
>The issue is this:
>
>* When a service loadbalances to external IPs (a typical example is the kubernetes API server service), cilium will rewrite the destination IP to that external IP and then give the packet to Linux for routing. The source IP is the one of the container. The container IP is not routable if we are running in encapsulation mode
>
>* When talking to external IPs outside of the cluster, masquerading is required anyway. This can be done by the cloud provider in which case users would run with masquerading turned off.

But on both cases, doesn't it bpf solves it by creating a SNAT with a bpf map to keep the relation between container:port <-> hostIP:port and snoop on all interfaces to translate it back?

**Comment by tgraf** [comment]
> But on both cases, doesn't it bpf solves it by creating a SNAT with a bpf map to keep the relation between container:port <-> hostIP:port and snoop on all interfaces to translate it back?

We would need full PAT with port allocation as we will have clashing 5-tuples. Not something that is very trivial. I agree that we should do it in BPF once we have figured out how to do it.

**Pull request merged**
